go-controller/pkg/util/gateway-init.go

```diff
@@ -39,68 +39,80 @@ func getLocalSystemID() (string, error) {
 	return localSystemID, nil
 }
 
-func lockNBForGateways() error {
-	localSystemID, err := getLocalSystemID()
+// GetDefaultGatewayRouterIP returns the first gateway logical router name
+// and IP address as listed in the OVN database
+func GetDefaultGatewayRouterIP() (string, net.IP, error) {
+	stdout, stderr, err := RunOVNNbctl("--data=bare", "--format=table",
+		"--no-heading", "--columns=name,options", "find", "logical_router",
+		"options:lb_force_snat_ip!=-")
 	if err != nil {
-		return err
+		return "", nil, fmt.Errorf("failed to get logical routers, stdout: %q, "+
+			"stderr: %q, err: %v", stdout, stderr, err)
 	}
-
-	stdout, stderr, err := RunOVNNbctlWithTimeout(60, "--", "wait-until",
-		"nb_global", ".", "external-ids:gateway-lock=\"\"", "--", "set",
-		"nb_global", ".", "external_ids:gateway-lock="+localSystemID)
-	if err != nil {
-		return fmt.Errorf("Failed to set gateway-lock "+
-			"stdout: %q, stderr: %q, error: %v", stdout, stderr, err)
+	// Convert \r\n to \n to support Windows line endings
+	stdout = strings.Replace(strings.TrimSpace(stdout), "\r\n", "\n", -1)
+	gatewayRouters := strings.Split(stdout, "\n")
+	if len(gatewayRouters) == 0 {
+		return "", nil, fmt.Errorf("failed to get default gateway router (no routers found)")
 	}
-	return nil
+	parts := strings.Fields(gatewayRouters[0])
+	var ip net.IP
+	for _, p := range parts {
+		const forceTag string = "lb_force_snat_ip="
+		if strings.HasPrefix(p, forceTag) {
+			ip = net.ParseIP(p[len(forceTag):])
+			break
+		}
+	}
+	if ip == nil {
+		return "", nil, fmt.Errorf("failed to parse gateway router %q IP", parts[0])
+	}
+	return parts[0], ip, nil
 }
 
-func unlockNBForGateways() {
-	stdout, stderr, err := RunOVNNbctl("--", "set", "nb_global", ".",
-		"external-ids:gateway-lock=\"\"")
-	if err != nil {
-		logrus.Errorf("Failed to delete lock for gateways, "+
-			"stdout: %q, stderr: %q, error: %v", stdout, stderr, err)
+func ensureGatewayPortAddress(portName string) (string, *net.IPNet, error) {
+	routerMac, routerIP, _ := GetPortAddresses(portName, false)
+	if routerMac == "" || routerIP == "" {
+		// Create the gateway switch port in 'join' if it doesn't exist yet
+		stdout, stderr, err := RunOVNNbctl("--wait=sb",
+			"--may-exist", "lsp-add", "join", portName,
+			"--", "--if-exists", "clear", "logical_switch_port", portName, "dynamic_addresses",
+			"--", "lsp-set-addresses", portName, "dynamic")
+		if err != nil {
+			return "", nil, fmt.Errorf("failed to add logical switch "+
+				"port %s, stdout: %q, stderr: %q, error: %v",
+				portName, stdout, stderr, err)
+		}
+		// Wait for its addresses
+		routerMac, routerIP, err = GetPortAddresses(portName, false)
+		if err != nil {
+			return "", nil, fmt.Errorf("Error while waiting for addresses "+
+				"for gateway switch port %q: %v", portName, err)
+		}
 	}
```

Comment:
Here err == nil doesn't necessarily mean routerMac and routerIP is set, right? GetPortAddresses returns nil with routerMac and routerIP set to ""